run/john.conf

```diff
@@ -4008,18 +4003,24 @@ void new()
 		rotor_cnt[wlen] = rotor_idx[wlen] = 0;
 		rotor_ptr[wlen] = rotor_off;
 		ch = word[wlen];
-		if (ch >= 'A' && ch <= 'Z') {
-			ch += 0x20;
-			word[wlen] = ch;
-			rotor[rotor_off++] = ch;
-			rotor[rotor_off++] = ch-0x20;
-		}
-		if (ch >= 'a' && ch <= 'z') {
-			rotor[rotor_off++] = ch;
-			rotor[rotor_off++] = ch-0x20;
-			rotor_cnt[wlen] = 2;
-			hybrid_total *= 2;
-			rotors[idx++] = wlen;
+		/* traditionally, this block was always executed, with Wordcase
+		   mode added we execute it either always when word_mode isn't
+		   used or at the beginning of a word or when the previous char
+		   was space */
+		if (word_mode == 0 || wlen == 0 || word[wlen-1] == 0x20) {
```

Comment:
this is the new condition, however it relies on short-circuit evaluation assuming that when wlen == 0 the test word[wlen-1] == 0x20 is skipped; so this might introduce an out-of-bounds memory access. any ideas how to best fix this ?

Reply:
Conservative changes just to this place:

```c
int prevch;
prevch = 0x20;
if (wlen) prevch = word[wlen  - 1];
if (!word_mode || prevch == 0x20) {
```

A better fix would probably be to initialize prevch out of the loop and to add `prevch = ch;` near the `++wlen;` line.

BTW, `!word_mode` is a bit quicker than `word_mode == 0`. (The external mode compiler is mostly non-optimizing.)

Reply:
thanks! pushed a new commit addressing this, i made it into a separate commit because i thought this might be interesting for other log readers, but feel free to squash as desired.